### PR TITLE
Keystore refactor to enable GTests mockings

### LINKED ISSUE #21757: Keystore refactor to enable GTests mockings
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.8.0 | Wazuh component | Manager | Sources | Linux |

## Description
As part of the process to incorporate the `Keystore` app (wazuh-keystore) ref. #21573, and to allow mocking `OpenSSL` `C` functions using `GTest`, it is necessary to refactor/template part of the code to follow the guidelines used in these cases where `C++` code coexists with calls to `C` functions.

### PR DESCRIPTION
**Pull request by tdrauncieal**
|Related issue|
|---|
|Closes #21757|

## Description
As part of the process to incorporate the `Keystore` app (wazuh-keystore) ref. #21573, and to allow mocking `OpenSSL` `C` functions using `GTest`, it is necessary to refactor/template part of the code to follow the guidelines used in these cases where `C++` code coexists with calls to `C` functions.

## Tests
```shell
[==========] Running 13 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 13 tests from OpenSSLWrapperTest
[ RUN      ] OpenSSLWrapperTest.OpenSSLWrapperTestInstance
[       OK ] OpenSSLWrapperTest.OpenSSLWrapperTestInstance (0 ms)
[ RUN      ] OpenSSLWrapperTest.getPubKeyFromCertMissingCertficate
[       OK ] OpenSSLWrapperTest.getPubKeyFromCertMissingCertficate (0 ms)
[ RUN      ] OpenSSLWrapperTest.getPubKeyFromCertMissingPublicKey
[       OK ] OpenSSLWrapperTest.getPubKeyFromCertMissingPublicKey (0 ms)
[ RUN      ] OpenSSLWrapperTest.createRSAInvalidRSAFile
[       OK ] OpenSSLWrapperTest.createRSAInvalidRSAFile (0 ms)
[ RUN      ] OpenSSLWrapperTest.createRSAInvalidPrivateKey
[       OK ] OpenSSLWrapperTest.createRSAInvalidPrivateKey (0 ms)
[ RUN      ] OpenSSLWrapperTest.createRSAInvalidPublicKey
[       OK ] OpenSSLWrapperTest.createRSAInvalidPublicKey (0 ms)
[ RUN      ] OpenSSLWrapperTest.rsaEncryptFailed
[       OK ] OpenSSLWrapperTest.rsaEncryptFailed (0 ms)
[ RUN      ] OpenSSLWrapperTest.rsaEncryptRSACertUnsuportedType
[       OK ] OpenSSLWrapperTest.rsaEncryptRSACertUnsuportedType (0 ms)
[ RUN      ] OpenSSLWrapperTest.rsaEncryptRSACertErrorExtracting
[       OK ] OpenSSLWrapperTest.rsaEncryptRSACertErrorExtracting (0 ms)
[ RUN      ] OpenSSLWrapperTest.rsaEncryptRSAPublicSucces
[       OK ] OpenSSLWrapperTest.rsaEncryptRSAPublicSucces (0 ms)
[ RUN      ] OpenSSLWrapperTest.rsaEncryptRSACertSucces
[       OK ] OpenSSLWrapperTest.rsaEncryptRSACertSucces (0 ms)
[ RUN      ] OpenSSLWrapperTest.rsaDecryptDecryptionFailed
[       OK ] OpenSSLWrapperTest.rsaDecryptDecryptionFailed (0 ms)
[ RUN      ] OpenSSLWrapperTest.rsaDecryptSuccess
[       OK ] OpenSSLWrapperTest.rsaDecryptSuccess (0 ms)
[----------] 13 tests from OpenSSLWrapperTest (2 ms total)

[----------] Global test environment tear-down
[==========] 13 tests from 1 test suite ran. (2 ms total)
[  PASSED  ] 13 tests.
``` 

- Compilation without warnings in every supported platform
  - [X] Linux
  - [ ] Windows
  - [ ] MAC OS X
- [X] Source installation
- [ ] Package installation
- [ ] Source upgrade
- [ ] Package upgrade
- [ ] Review logs syntax and correct language
- [ ] QA templates contemplate the added capabilities

### CODE REVIEW

**Comment by tdrauncieal** [comment]
Due to the failed check related:
```shell
macOS compilation test / build-ventura (pull_request) Failing after 46s
``` 
the following issue has been created:
[MacOS 13 linker parameter fail #21830](https://github.com/wazuh/wazuh/issues/21830)

File: src/shared_modules/keystore/include/keyStore.hpp [hunk 1]
```diff
@@ -14,23 +14,52 @@
...
 
+#include "loggerHelper.h"
+#include "opensslWrapper.hpp"
 #include "rocksDBWrapper.hpp"
 
 constexpr auto DATABASE_PATH {"queue/keystore"};
 constexpr auto PRIVATE_KEY_FILE {"etc/sslmanager.key"};
 constexpr auto CERTIFICATE_FILE {"etc/sslmanager.cert"};
 
+constexpr auto KS_NAME {"keystore"};
```
Comment (ncvicchi) [inline src/shared_modules/keystore/include/keyStore.hpp:25]: No sería mejor KEYSTORE_NAME?

**Review by ncvicchi — APPROVED** [review]

File: src/shared_modules/keystore/include/keyStore.hpp (hunk 1 above)
Comment (tdrauncieal) [reply src/shared_modules/keystore/include/keyStore.hpp:25]: this code was imported from eliminated file `rsaHelper.hpp` and was reviewed in that time, But if you consider this change appropriate, I will proceed.

Comment (mateocervilla) [reply src/shared_modules/keystore/include/keyStore.hpp:25]: When creating the constant I used the same terminology as the `indexer-connector`
```cpp
constexpr auto IC_NAME {"indexer-connector"};
```

**Review by mateocervilla — APPROVED** [review]
LGTM!

File: src/shared_modules/keystore/include/keyStore.hpp (hunk 1 above)
Comment (ncvicchi) [reply src/shared_modules/keystore/include/keyStore.hpp:25]: Let's keep it then. I was just curious.

**Comment by ncvicchi** [comment]
LGTM!

File: src/shared_modules/utils/opensslWrapper.hpp [hunk 2]
```diff
@@ -24,41 +25,130 @@ constexpr int RSA_PRIVATE {0};
...
+     * @return          The size of the encrypted output, -1 if error
+     */
+    int rsaEncrypt(const std::string& filePath, const std::string& input, std::string& output, bool cert = false)
+    {
+        RSA* rsa = nullptr;
+
+        createRSA(rsa, filePath, cert ? RSA_CERT : RSA_PUBLIC);
+
+        // Allocate memory for the encryptedValue
+        unsigned char* encryptedValue = (unsigned char*)malloc(T::RSA_size(rsa));
```
Comment (Dwordcito) [inline src/shared_modules/utils/opensslWrapper.hpp:53]: Why are we using the old implementation? The previous function was different and did not use malloc.

```cpp
    int rsaEncrypt(const std::string& filePath, const std::string& input, std::string& output, bool cert = false)
    {
        RSA* rsa = nullptr;

        createRSA(rsa, filePath, cert ? RSA_CERT : RSA_PUBLIC);

        // Allocate memory for the encryptedValue
        std::vector<unsigned char> encryptedValue(RSA_size(rsa));

        // Defered free
        DEFER([&]() { T::RSA_free(rsa); });

        const auto encryptedLen = T::RSA_public_encrypt(
            input.length(), (const unsigned char*)input.data(), encryptedValue.data(), rsa, RSA_PKCS1_PADDING);

        if (encryptedLen < 0)
        {
            throw std::runtime_error("RSA encryption failed");
        }

        output = std::string(encryptedValue.begin(), encryptedValue.end());

        return encryptedLen;
    }

```

File: src/shared_modules/utils/opensslWrapper.hpp [hunk 3]
```diff
@@ -24,41 +25,130 @@ constexpr int RSA_PRIVATE {0};
...
 constexpr int RSA_CERT {2};
 
-namespace Utils
+template<typename T = OpenSSLPrimitives, typename U = OSPrimitives>
+class OpenSSL final
+    : public T
+    , public U
 {
+public:
+    explicit OpenSSL() {}
```
Comment (Dwordcito) [inline src/shared_modules/utils/opensslWrapper.hpp:34]:
```suggestion
    explicit OpenSSL() = default;
```

File: src/shared_modules/utils/opensslWrapper.hpp [hunk 4]
```diff
@@ -24,41 +25,130 @@ constexpr int RSA_PRIVATE {0};
...
+    virtual ~OpenSSL() {}
```
Comment (Dwordcito) [inline src/shared_modules/utils/opensslWrapper.hpp:35]:
```suggestion
    virtual ~OpenSSL() = default;
```

File: src/shared_modules/keystore/src/main.cpp [hunk 5]
```diff
@@ -12,6 +12,7 @@
 #include "argsParser.hpp"
 #include "homedirHelper.hpp"
 #include "keyStore.hpp"
+#include "opensslWrapper.hpp"
```
Comment (Dwordcito) [inline src/shared_modules/keystore/src/main.cpp:15]: not needed.

File: src/shared_modules/keystore/include/keyStore.hpp [hunk 6]
```diff
@@ -14,23 +14,52 @@
 
 #include <string>
 
+#include "loggerHelper.h"
+#include "opensslWrapper.hpp"
```
Comment (Dwordcito) [inline src/shared_modules/keystore/include/keyStore.hpp:18]: not needed

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp [hunk 7]
```diff
@@ -80,8 +81,8 @@ IndexerConnector::IndexerConnector(
         }
     }
 
-    Keystore::get(INDEXER_COLUMN, USER_KEY, username);
-    Keystore::get(INDEXER_COLUMN, PASSWORD_KEY, password);
+    Keystore<>::get(INDEXER_COLUMN, USER_KEY, username);
```
Comment (Dwordcito) [inline src/shared_modules/indexer_connector/src/indexerConnector.cpp:84]:
```suggestion
    Keystore::get(INDEXER_COLUMN, USER_KEY, username);
```

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp [hunk 8]
```diff
@@ -80,8 +81,8 @@ IndexerConnector::IndexerConnector(
...
+    Keystore<>::get(INDEXER_COLUMN, PASSWORD_KEY, password);
```
Comment (Dwordcito) [inline src/shared_modules/indexer_connector/src/indexerConnector.cpp:85]:
```suggestion
    Keystore::get(INDEXER_COLUMN, PASSWORD_KEY, password);
```

File: src/shared_modules/indexer_connector/src/indexerConnector.cpp [hunk 9]
```diff
@@ -13,6 +13,7 @@
 #include "HTTPRequest.hpp"
 #include "keyStore.hpp"
 #include "loggerHelper.h"
+#include "opensslWrapper.hpp"
```
Comment (Dwordcito) [inline src/shared_modules/indexer_connector/src/indexerConnector.cpp:16]: not needed

File: src/shared_modules/keystore/include/keyStore.hpp [hunk 10]
```diff
@@ -39,7 +68,32 @@ class Keystore final
...
+            }
+            // Decrypt value
+            m_openssl->rsaDecrypt(PRIVATE_KEY_FILE, encryptedValue, value);
+        }
+    }
 };
 
+template<class TOSSLPrimitive>
+std::shared_ptr<TOSSLPrimitive> Keystore<TOSSLPrimitive>::m_openssl {std::make_shared<TOSSLPrimitive>()};
+
```
Comment (Dwordcito) [inline src/shared_modules/keystore/include/keyStore.hpp:98]: class with static methods with static member, this not make sense.

```suggestion
using Keystore = TKeystore<>;

```

File: src/shared_modules/keystore/include/keyStore.hpp [hunk 11]
```diff
@@ -14,23 +14,52 @@
...
 
 constexpr auto DATABASE_PATH {"queue/keystore"};
 constexpr auto PRIVATE_KEY_FILE {"etc/sslmanager.key"};
 constexpr auto CERTIFICATE_FILE {"etc/sslmanager.cert"};
 
+constexpr auto KS_NAME {"keystore"};
+template<typename TOSSLPrimitive = OpenSSL<>>
 class Keystore final
 {
+    static std::shared_ptr<TOSSLPrimitive> m_openssl;
```
Comment (Dwordcito) [inline src/shared_modules/keystore/include/keyStore.hpp:29]:
```suggestion
```

File: src/shared_modules/keystore/include/keyStore.hpp [hunk 12]
```diff
@@ -39,7 +68,32 @@ class Keystore final
...
+
+        if (keystoreDB.get(key, encryptedValue, columnFamily))
+        {
+            if (!std::filesystem::exists(PRIVATE_KEY_FILE))
+            {
+                logWarn(KS_NAME, "No private key was found.");
+                return;
+            }
+            // Decrypt value
+            m_openssl->rsaDecrypt(PRIVATE_KEY_FILE, encryptedValue, value);
```
Comment (Dwordcito) [inline src/shared_modules/keystore/include/keyStore.hpp:91]:
```suggestion
            TOSSLPrimitive().rsaDecrypt(PRIVATE_KEY_FILE, encryptedValue, value);
```

File: src/shared_modules/keystore/include/keyStore.hpp [hunk 13]
```diff
@@ -14,23 +14,52 @@
...
+    {
+        std::string encryptedValue;
+
+        if (!std::filesystem::exists(CERTIFICATE_FILE))
+        {
+            logWarn(KS_NAME, "No certificate was found.");
+            return;
+        }
+        // Encrypt value
+        m_openssl->rsaEncrypt(CERTIFICATE_FILE, value, encryptedValue, true);
```
Comment (Dwordcito) [inline src/shared_modules/keystore/include/keyStore.hpp:51]:
```suggestion
        TOSSLPrimitive().rsaEncrypt(CERTIFICATE_FILE, value, encryptedValue, true);
```

File: src/shared_modules/keystore/src/main.cpp [hunk 14]
```diff
@@ -39,7 +40,7 @@ int main(int argc, char* argv[])
         key = args.getKey();
         value = args.getValue();
 
-        Keystore::put(family, key, value);
+        Keystore<>::put(family, key, value);
```
Comment (Dwordcito) [inline src/shared_modules/keystore/src/main.cpp:43]:
```suggestion
        Keystore::put(family, key, value);
```

File: src/shared_modules/utils/opensslWrapper.hpp [hunk 15]
```diff
@@ -1,21 +1,22 @@
 /*
  * Wazuh shared modules utils
  * Copyright (C) 2015, Wazuh Inc.
- * January 24, 2024.
+ * Febrary 05, 2024.
```
Comment (Dwordcito) [inline src/shared_modules/utils/opensslWrapper.hpp:4]: not change the date if the file suffers some changes

File: src/shared_modules/utils/opensslWrapper.hpp
Comment (Dwordcito) [inline src/shared_modules/utils/opensslWrapper.hpp:1]: I don't think we should change the name, this is a helper.

File: src/shared_modules/utils/opensslWrapper.hpp [hunk 16]
```diff
@@ -24,41 +25,130 @@ constexpr int RSA_PRIVATE {0};
...
+
+        // Allocate memory for the encryptedValue
+        unsigned char* encryptedValue = (unsigned char*)malloc(T::RSA_size(rsa));
+
+        // Defered free
+        DEFER(
+            [&]()
+            {
+                T::RSA_free(rsa);
+                free(encryptedValue);
```
Comment (Dwordcito) [inline src/shared_modules/utils/opensslWrapper.hpp:60]: avoid malloc/free in cpp.

File: src/shared_modules/utils/opensslPrimitives.hpp [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+
+#include <openssl/pem.h>
+#include <openssl/rsa.h>
+
+class OpenSSLPrimitives
+{
+protected:
+    OpenSSLPrimitives() = default;
+    virtual ~OpenSSLPrimitives() = default;
+
```
Comment (Dwordcito) [inline src/shared_modules/utils/opensslPrimitives.hpp:23]:
```suggestion
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
```

File: src/shared_modules/utils/opensslPrimitives.hpp [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+    inline rsa_st* EVP_PKEY_get1_RSA(EVP_PKEY* pkey)
+    {
+        return ::EVP_PKEY_get1_RSA(pkey);
+    }
+
+    inline int EVP_PKEY_get_base_id(const EVP_PKEY* pkey)
+    {
+        return ::EVP_PKEY_get_base_id(pkey);
+    }
```
Comment (Dwordcito) [inline src/shared_modules/utils/opensslPrimitives.hpp:82]:
```suggestion
    }
    #pragma GCC diagnostic pop
```

**Review by Dwordcito — CHANGES_REQUESTED** [review]

File: src/shared_modules/utils/opensslWrapper.hpp
Comment (ncvicchi) [reply src/shared_modules/utils/opensslWrapper.hpp:1]: What do you think if we change but to opensslHelper.hpp instead?

Comment (Dwordcito) [reply src/shared_modules/utils/opensslWrapper.hpp:1]: This is a helper for RSA only, which is a subset of openssl's capabilities.

rsaHelper.hpp its the expected name.

Comment (ncvicchi) [reply src/shared_modules/utils/opensslWrapper.hpp:1]: Will do

File: src/shared_modules/utils/opensslPrimitives.hpp [hunk 19]
```diff
@@ -0,0 +1,89 @@
+/*
+ * Wazuh shared modules utils
+ * Copyright (C) 2015, Wazuh Inc.
+ * Febrary 05, 2024.
```
Comment (Dwordcito) [inline src/shared_modules/utils/opensslPrimitives.hpp:4]:
```suggestion
 * February 05, 2024.
```

File: src/shared_modules/utils/rsaHelper.hpp [hunk 20]
```diff
@@ -24,41 +25,125 @@ constexpr int RSA_PRIVATE {0};
...
+     * @return          The size of the encrypted output, -1 if error
+     */
+    int rsaEncrypt(const std::string& filePath, const std::string& input, std::string& output, bool cert = false)
+    {
+        RSA* rsa = nullptr;
+
+        createRSA(rsa, filePath, cert ? RSA_CERT : RSA_PUBLIC);
+
+        // Allocate memory for the encryptedValue
+        auto encryptedValue {std::make_unique<unsigned char[]>(T::RSA_size(rsa))};
```
Comment (Dwordcito) [inline src/shared_modules/utils/rsaHelper.hpp:53]: Again, use std::vector with size reserve.

File: src/shared_modules/utils/rsaHelper.hpp [hunk 21]
```diff
@@ -24,41 +25,125 @@ constexpr int RSA_PRIVATE {0};
...
+
+        std::string decryptedText(T::RSA_size(rsa), 0); // Initialize with zeros
+
+        // Defered free
+        DEFER([&]() { T::RSA_free(rsa); });
+
+        // Decrypt the ciphertext using RSA private key
+        const auto decryptedLen = T::RSA_private_decrypt(256,
+                                                         reinterpret_cast<const unsigned char*>(input.data()),
+                                                         reinterpret_cast<unsigned char*>(&decryptedText[0]),
```
Comment (Dwordcito) [inline src/shared_modules/utils/rsaHelper.hpp:94]: avoid index access. use const_cast with data()

File: src/shared_modules/utils/rsaHelper.hpp [hunk 22]
```diff
@@ -24,41 +25,125 @@ constexpr int RSA_PRIVATE {0};
...
+        createRSA(rsa, filePath, cert ? RSA_CERT : RSA_PUBLIC);
+
+        // Allocate memory for the encryptedValue
+        auto encryptedValue {std::make_unique<unsigned char[]>(T::RSA_size(rsa))};
+
+        // Defered free
+        DEFER([&]() { T::RSA_free(rsa); });
+
+        const auto encryptedLen = T::RSA_public_encrypt(
+            input.length(), (const unsigned char*)input.data(), &encryptedValue[0], rsa, RSA_PKCS1_PADDING);
```
Comment (Dwordcito) [inline src/shared_modules/utils/rsaHelper.hpp:59]: avoid c cast.

File: src/shared_modules/utils/rsaHelper.hpp [hunk 23]
```diff
@@ -24,41 +25,125 @@ constexpr int RSA_PRIVATE {0};
...
 
         // Defered free
-        DEFER([x509Certificate]() { X509_free(x509Certificate); });
+        DEFER([&]() { T::X509_free(x509Certificate); });
 
         // Extract the public key from the X.509 certificate
-        EVP_PKEY* evpPublicKey = X509_get_pubkey(x509Certificate);
+        EVP_PKEY* evpPublicKey = T::X509_get_pubkey(x509Certificate);
+
+        DEFER([&]() { T::EVP_PKEY_free(evpPublicKey); });
```
Comment (Dwordcito) [inline src/shared_modules/utils/rsaHelper.hpp:135]: move this line to after the `if (!evpPublicKey)`

**Review by Dwordcito — CHANGES_REQUESTED** [review]

**Review by Dwordcito — APPROVED** [review]

**Pull request merged**